llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -14112,7 +14102,9 @@ static SDValue combineBinOp_VLToVWBinOp_VL(SDNode *N,
                                            TargetLowering::DAGCombinerInfo &DCI,
                                            const RISCVSubtarget &Subtarget) {
   SelectionDAG &DAG = DCI.DAG;
-  if (DCI.isBeforeLegalize())
+  // Don't perform this until types are legalized and any legal i1 types are
+  // custom lowered to avoid introducing unselectable V{S,Z}EXT_VLs.
+  if (DCI.isBeforeLegalizeOps())
```

Comment:
I'm not sure this is 100% reliable. Its theoretically possible for an i1 vector to be created by the DAG combiner after legalize ops. The last DAG combine stage also runs the legalizer on every node as  part of its worklist. So its not illegal for an i1 zext to created as it would get legalized before isel.

Reply:
Maybe it is better for us to still check whether the op is legal. 
But as @lukel97 said, the original EEW / 2 check, which the VP intrinsics already does, could be removed, AFAIU ?

Reply:
I've updated the PR to instead check that the narrow element type isn't i1 across the different possible extend ops